samples/opengl/depthMapTriangulation/image_triangulator.cpp

```diff
@@ -0,0 +1,949 @@
+            quad_heap.push_back(
+                std::tuple<Point2D, int, bool>(Point2D(curr_y, curr_x + step_mid), next_side_len, true));
+            child_4 = true;
+        }
+        else
+        {
+            child_4 = false;
+        }
+
+        // do the checking for all the triangles not added to the heap.
+        // there can be 0 - 2 such unhandled triangles for all 4 directions
+
+        // check if the top needs 0, 1 or 2 triangles
+        if (!up_exists)
+        {
+            add_face_to_mesh(current_mesh, vh1, vhc, vh4);
+        }
+        else
+        {
+            // triangulate current_mesh respective triangles if they are not handled by a recursive call
+            if (!child_1)
+                add_face_to_mesh(current_mesh, vh1, vhc, selected_vertices(curr_y, curr_x + step_mid));  // top left
+            if (!child_4)
+                add_face_to_mesh(current_mesh, vhc, vh4, selected_vertices(curr_y, curr_x + step_mid));  // top right
+        }
+
+        // also check the left
+        if (!left_exists)
+        {
+            add_face_to_mesh(current_mesh, vh1, vh2, vhc);
+        }
+        else
+        {
+            if (!child_1)
+                add_face_to_mesh(current_mesh, vhc, vh1, selected_vertices(curr_y + step_mid, curr_x));  // left upper
+            if (!child_2)
+                add_face_to_mesh(current_mesh, vh2, vhc, selected_vertices(curr_y + step_mid, curr_x));  // left lower
+        }
+
+        // also check below
+        if (!down_exists)
+        {
+            add_face_to_mesh(current_mesh, vhc, vh2, vh3);
+        }
+        else
+        {
+            if (!child_2)
+                add_face_to_mesh(current_mesh, vhc, vh2,
+                                 selected_vertices(curr_max_y, curr_x + step_mid));  // bottom left
+            if (!child_3)
+                add_face_to_mesh(current_mesh, vh3, vhc,
+                                 selected_vertices(curr_max_y, curr_x + step_mid));  // bottom right
+        }
+
+        // also check the right
+        if (!right_exists)
+        {
+            add_face_to_mesh(current_mesh, vhc, vh3, vh4);
+        }
+        else
+        {
+            if (!child_3)
+                add_face_to_mesh(current_mesh, vhc, vh3,
+                                 selected_vertices(curr_y + step_mid, curr_max_x));  // right lower
+            if (!child_4)
+                add_face_to_mesh(current_mesh, vh4, vhc,
+                                 selected_vertices(curr_y + step_mid, curr_max_x));  // right upper
+        }
+    }
+}
+
+
+// ----------- error calculation stuff -----------
+
+bool Saiga::RQT_Triangulator::check_metric(OpenMesh::Vec3f const& p, OpenMesh::Vec3f const& triangle_vertex,
+                                           OpenMesh::Vec3f const& normal, float threshold)
+{
+    if (abs(p[2]) < 0.00005f) return true;
+
+    float square_dist = pow(dot(normal, p - triangle_vertex), 2) / dot(normal, normal);
```

Comment:
Never use `pow(x,2)`. This is incredible slow.

I also don't quite get why you're dividing by `dot(normal,normal)`. Is the normal not normalized?